src/item.cpp

```diff
@@ -7517,6 +7544,7 @@ void item::process_temperature_rot( float insulation, const tripoint &pos,
 
                 if( has_rotten_away() || ( is_corpse() && rot > 10_days ) ) {
                     // No need to track item that will be gone
+                    item_internal::goes_bad_cache_unset();
```

Comment:
Rather than having to remember to unset this on each exit, it would be better to use an RAII helper.  That's a struct which sets up the cache in its constructor and clears it out in its destructor.  The destructor will automatically get called on all function exits (including ones you might not be thinking about like via thrown exceptions).